Sources/PriorityQueueModule/PriorityQueue.swift

```diff
@@ -0,0 +1,363 @@
+                }
+
+                _trickleDownMin(startingAt: smallestDescendantIdx)
+            }
+        }
+    }
+
+    private mutating func _trickleDownMax(startingAt index: Int) {
+        let leftIdx = _leftChildIndex(of: index)
+
+        guard leftIdx < storage.count else {
+            // We have no descendants -- no need to trickle down further
+            return
+        }
+
+        guard let (largestDescendantIdx, isChild) = _indexOfChildOrGrandchild(of: index, sortedUsing: >) else {
+            return
+        }
+
+        if isChild {
+            if storage[largestDescendantIdx] > storage[index] {
+                storage.swapAt(largestDescendantIdx, index)
+            }
+        } else {
+            // Largest is a grandchild
+            if storage[largestDescendantIdx] > storage[index] {
+                storage.swapAt(largestDescendantIdx, index)
+
+                let parentIdx = _parentIndex(of: largestDescendantIdx)
+                if storage[largestDescendantIdx] < storage[parentIdx] {
+                    storage.swapAt(largestDescendantIdx, parentIdx)
+                }
+
+                _trickleDownMax(startingAt: largestDescendantIdx)
+            }
+        }
+    }
+
+    /// Returns the smallest or largest child or grandchild of the element at the given index,
+    /// as determined by the predicate.
+    ///
+    /// - parameter index: The index of the item whose descendants should be compared.
+    /// - parameter predicate: Returns `true` if its first argument should be ordered before its second argument.
+    private func _indexOfChildOrGrandchild(
+        of index: Int,
+        sortedUsing predicate: (Element, Element) -> Bool
+    ) -> (index: Int, isChild: Bool)? {
+        let leftChildIdx = _leftChildIndex(of: index)
+        guard leftChildIdx < storage.count else {
+            return nil
+        }
+
+        var result: (index: Int, isChild: Bool) = (leftChildIdx, true)
+
+        let rightChildIdx = _rightChildIndex(of: index)
+        guard rightChildIdx < storage.count else {
+            return result
+        }
+
+        // Compare the two children
+        if predicate(storage[rightChildIdx], storage[leftChildIdx]) {
+            result.index = rightChildIdx
+        }
+
+        let firstGrandchildIdx = _firstGrandchildIndex(of: index)
+        guard firstGrandchildIdx < storage.count else {
+            return result
+        }
+
+        // Iterate through the grandchildren
+        for i in firstGrandchildIdx..._lastGrandchildIndex(of: index) {
+            guard i < storage.count else {
+                return result
+            }
+
+            if predicate(storage[i], storage[result.index]) {
+                result.index = i
+                result.isChild = false
+            }
+        }
+
+        return result
+    }
+
+    // MARK: - Helpers
+
+    @inline(__always)
+    private func _parentIndex(of index: Int) -> Int {
+        (index - 1) / 2
+    }
```

Comment:
@timvermeulen - that should only be required for the public interface, [module private/internal methods should inline with just `@inline(__always)`](https://forums.swift.org/t/when-should-both-inlinable-and-inline-always-be-used/37375/2) unless I've missed something (and looking at the stack traces from Instruments, it definitely seems inlining have worked..). @inlinable would be required for the public methods though!

Reply:
Can somebody confirm this? Is `@inlinable` required?

Reply:
Private functions marked with `@inline(__always)` can definitely be inlined within the module, but it prevents the function it's inlined into from itself being inlined across module boundaries. You can easily see this by e.g. marking `popMin` with `@inlinable` — the compiler will complain that this is impossible because `_remove`, which `popMin` is implemented in terms of, is private and can therefore not be referenced from an `@inlinable` function. Hence my earlier suggestion to mark everything as `@inlinable` because the compiler will guide you to make this possible 🙂